File: ts-lib/src/api/tx-transfer.ts
import express from "express";
import {
  populateProvedTransfer,
  gasEstimateForUnprovenTransfer,
  generateTransferProof,
} from "dop-wallet-v3";

export const txTransferRouter = express.Router();

function safeJsonResponse(res: express.Response, data: any) {
  res.setHeader("Content-Type", "application/json");
  res.send(
    JSON.stringify(data, (key, value) =>
      typeof value === "bigint" ? value.toString() : value
    )
  );
}

// POST /wallet/populate-proved-transfer
txTransferRouter.post("/populate-proved-transfer", async (req, res) => {
  try {
    const {
      txidVersion,
      networkName,
      dopWalletID,
      showSenderAddressToRecipient,
      memoText,
      erc20AmountRecipients,
      nftAmountRecipients,
      broadcasterFeeERC20AmountRecipient,
      sendWithPublicWallet,
      overallBatchMinGasPrice,
      gasDetails,
    } = req.body;

    if (!txidVersion || !networkName || !dopWalletID || !gasDetails) {
      res.status(400).send("Missing required fields");
      return;
    }

    const tx = await populateProvedTransfer(
      txidVersion,
      networkName,
      dopWalletID,
      showSenderAddressToRecipient ?? false,
      memoText,
      erc20AmountRecipients ?? [],
      nftAmountRecipients ?? [],
      broadcasterFeeERC20AmountRecipient,
      sendWithPublicWallet ?? false,
      overallBatchMinGasPrice,
      gasDetails
    );

    safeJsonResponse(res, tx);
  } catch (err) {
    console.error("❌ Failed to populate proved transfer:", err);
    res.status(500).json({
      error: "Failed to populate proved transfer",
      details: (err as Error)?.message ?? String(err),
    });
  }
});

// POST /wallet/gas-estimate-for-unproven-transfer
txTransferRouter.post(
  "/gas-estimate-for-unproven-transfer",
  async (req, res) => {
    try {
      const {
        txidVersion,
        networkName,
        dopWalletID,
        encryptionKey,
        memoText,
        erc20AmountRecipients,
        nftAmountRecipients,
        originalGasDetails,
        feeTokenDetails,
        sendWithPublicWallet,
      } = req.body;

      if (
        !txidVersion ||
        !networkName ||
        !dopWalletID ||
        !encryptionKey ||
        !originalGasDetails
      ) {
        res.status(400).send("Missing required fields");
        return;
      }

      const estimate = await gasEstimateForUnprovenTransfer(
        txidVersion,
        networkName,
        dopWalletID,
        encryptionKey,
        memoText,
        erc20AmountRecipients ?? [],
        nftAmountRecipients ?? [],
        originalGasDetails,
        feeTokenDetails,
        sendWithPublicWallet ?? false
      );

      safeJsonResponse(res, estimate);
    } catch (err) {
      console.error("❌ Failed to estimate gas for unproven transfer:", err);
      res.status(500).json({
        error: "Failed to estimate gas for unproven transfer",
        details: (err as Error)?.message ?? String(err),
      });
    }
  }
);

txTransferRouter.post("/generate-transfer-proof", async (req, res) => {
  try {
    const {
      txidVersion,
      networkName,
      dopWalletID,
      encryptionKey,
      showSenderAddressToRecipient,
      memoText,
      erc20AmountRecipients,
      nftAmountRecipients,
      broadcasterFeeERC20AmountRecipient,
      sendWithPublicWallet,
      overallBatchMinGasPrice,
    } = req.body;

    if (!txidVersion || !networkName || !dopWalletID || !encryptionKey) {
      res.status(400).send("Missing required fields");
      return;
    }

    await generateTransferProof(
      txidVersion,
      networkName,
      dopWalletID,
      encryptionKey,
      showSenderAddressToRecipient ?? false,
      memoText,
      erc20AmountRecipients ?? [],
      nftAmountRecipients ?? [],
      broadcasterFeeERC20AmountRecipient,
      sendWithPublicWallet ?? false,
      overallBatchMinGasPrice,
      (_progress) => {
        console.log("🔄 Generating transfer proof...");
      }
    );

    res.status(200).json({ message: "Transfer proof generation complete" });
  } catch (err) {
    console.error("❌ Failed to generate transfer proof:", err);
    res.status(500).json({
      error: "Failed to generate transfer proof",
      details: (err as Error)?.message ?? String(err),
    });
  }
});

File: src/dop/tx_transfer.rs
use super::{
    DopClient, DopERC20AmountRecipient, DopNFTAmountRecipient,
    DopPopulateTransactionResponse, DopTransactionGasEstimateResponse, FeeTokenDetails,
    TransactionGasDetails,
};
use serde_json::json;

impl DopClient {
    pub async fn populate_proved_transfer(
        &self,
        txid_version: String,
        network_name: String,
        dop_wallet_id: String,
        show_sender_address_to_recipient: bool,
        memo_text: Option<String>,
        erc20_recipients: Vec<DopERC20AmountRecipient>,
        nft_recipients: Vec<DopNFTAmountRecipient>,
        broadcaster_fee_recipient: Option<DopERC20AmountRecipient>,
        send_with_public_wallet: bool,
        overall_batch_min_gas_price: Option<String>,
        gas_details: TransactionGasDetails,
    ) -> anyhow::Result<DopPopulateTransactionResponse> {
        let gas_details_json = serde_json::to_value(gas_details)?;

        let payload = serde_json::json!({
            "txidVersion": txid_version,
            "networkName": network_name,
            "dopWalletID": dop_wallet_id,
            "showSenderAddressToRecipient": show_sender_address_to_recipient,
            "memoText": memo_text,
            "erc20AmountRecipients": erc20_recipients,
            "nftAmountRecipients": nft_recipients,
            "broadcasterFeeERC20AmountRecipient": broadcaster_fee_recipient,
            "sendWithPublicWallet": send_with_public_wallet,
            "overallBatchMinGasPrice": overall_batch_min_gas_price,
            "gasDetails": gas_details_json,
        });

        let resp = self
            .client
            .post(&format!("{}/populate-proved-transfer", self.base_url()))
            .json(&payload)
            .send()
            .await?;

        println!("Response status: {:?}", resp.status());
        let status = resp.status();
        if !status.is_success() {
            let text = resp.text().await.unwrap_or_default();
            return Err(anyhow::anyhow!(
                "API call failed: {} (status {})",
                text,
                status
            ));
        }

        let result = resp.json::<DopPopulateTransactionResponse>().await?;
        Ok(result)
    }

    pub async fn gas_estimate_for_unproven_transfer(
        &self,
        txid_version: String,
        network_name: String,
        dop_wallet_id: String,
        encryption_key: String,
        memo_text: Option<String>,
        erc20_recipients: Vec<DopERC20AmountRecipient>,
        nft_recipients: Vec<DopNFTAmountRecipient>,
        original_gas_details: TransactionGasDetails,
        fee_token_details: Option<FeeTokenDetails>,
        send_with_public_wallet: bool,
    ) -> anyhow::Result<DopTransactionGasEstimateResponse> {
        let gas_details_json = serde_json::to_value(original_gas_details)?;

        let payload = serde_json::json!({
            "txidVersion": txid_version,
            "networkName": network_name,
            "dopWalletID": dop_wallet_id,
            "encryptionKey": encryption_key,
            "memoText": memo_text,
            "erc20AmountRecipients": erc20_recipients,
            "nftAmountRecipients": nft_recipients,
            "originalGasDetails": gas_details_json,
            "feeTokenDetails": fee_token_details,
            "sendWithPublicWallet": send_with_public_wallet,
        });

        let resp = self
            .client
            .post(&format!(
                "{}/gas-estimate-for-unproven-transfer",
                self.base_url()
            ))
            .json(&payload)
            .send()
            .await?;

        println!("Response status: {:?}", resp.status());
        let status = resp.status();
        if !status.is_success() {
            let text = resp.text().await.unwrap_or_default();
            return Err(anyhow::anyhow!(
                "API call failed: {} (status {})",
                text,
                status
            ));
        }

        let result = resp.json::<DopTransactionGasEstimateResponse>().await?;
        Ok(result)
    }

    pub async fn generate_transfer_proof(
        &self,
        txid_version: String,
        network_name: String,
        dop_wallet_id: String,
        encryption_key: String,
        show_sender_address_to_recipient: bool,
        memo_text: Option<String>,
        erc20_recipients: Vec<DopERC20AmountRecipient>,
        nft_recipients: Vec<DopNFTAmountRecipient>,
        broadcaster_fee_recipient: Option<DopERC20AmountRecipient>,
        send_with_public_wallet: bool,
        overall_batch_min_gas_price: Option<String>,
    ) -> Result<(), anyhow::Error> {
        let payload = json!({
            "txidVersion": txid_version,
            "networkName": network_name,
            "dopWalletID": dop_wallet_id,
            "encryptionKey": encryption_key,
            "showSenderAddressToRecipient": show_sender_address_to_recipient,
            "memoText": memo_text,
            "erc20AmountRecipients": erc20_recipients,
            "nftAmountRecipients": nft_recipients,
            "broadcasterFeeERC20AmountRecipient": broadcaster_fee_recipient,
            "sendWithPublicWallet": send_with_public_wallet,
            "overallBatchMinGasPrice": overall_batch_min_gas_price,
        });

        let resp = self
            .client
            .post(&format!("{}/generate-transfer-proof", self.base_url()))
            .json(&payload)
            .send()
            .await?;

        println!("Response status: {:?}", resp.status());
        let status = resp.status();
        if !status.is_success() {
            let text = resp.text().await.unwrap_or_default();
            return Err(anyhow::anyhow!(
                "API call failed: {} (status {})",
                text,
                status
            ));
        }

        Ok(())
    }
}
